### PR TITLE
Initial proposal for extended bytestring builtins (SCP-1163)

### PR DESCRIPTION
**Pull request by kwxm**
This PR just adds a document about possible new bytestring operations, mostly for discussion purposes.  We may not want to actually merge it.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A pkgsLocal.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [ ] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [ ] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: notes/plutus-core/Bytestrings.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+Plutus Core also provides built-in character and string types.  These both have
+concrete syntactical representations: `(con char 'q')` and `(con string
+"something")`.  Support for these is very limited: we only have `charToString`
+and `append` (append one string to another); also the PlutusTx compiler can
+compile literal Haskell strings and characters into PLC values of the
+appropriate types (but see below for some complications involving Strings).
+
+However there is a subtlety: Plutus Core actually has _two_ representations of
+Haskell Strings.  The PlutusTx plugin compiles normal Haskell strings into
+Scott-encoded lists of PLC's `char` type ("Scott strings" for short), but there
```
Comment (michaelpj) [inline notes/plutus-core/Bytestrings.md:47]: Hmmm. I wonder if we could just stop doing that and refuse to support non-overloaded string literals :thinking:

File: notes/plutus-core/Bytestrings.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  * A `consChar` operation allowing you to prepend a character to a PLC string. This would help with
+    building bytestrings dynamically.  Also, `PlutusTx.String` contains
+    a `stringToBuiltinString` function which converts Haskell `String` to `Builtin.String` by
+    repeated application of `charToString` and `appendString`, which is not very efficient.  I think that
+   `consChar` would allow us to implement this a low more efficiently.
+  * Alternatively, a `consByte` operation directly adding a character or integer to the front of a bytestring.
+    `Data.Bytestring.Char8.cons` would be helpful for this.
+  * Perhaps also
+     * A `lengthOfBytestring` function
+     * An `indexBytestring` function allowing you to retrieve the character or integer at a given position in a bytestring.
```
Comment (michaelpj) [inline notes/plutus-core/Bytestrings.md:111]: I guess we have to be careful with these functions to make it clear that it's the `Char8` version of length/index.

File: notes/plutus-core/Bytestrings.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+Plutus Core bytestrings are implemented using Haskell's (strict) `ByteString`
+type from `Data.ByteString`.  Internally these are implemented as arrays of
+`Word8` values.  Haskell doesn't provide any concrete syntax for the
+`ByteString` type, but there are functions `pack :: [Word8] -> ByteString` and
+`unpack :: ByteString -> [Word8]` which convert back and forth between
+`ByteString`s and lists of `Word8` values.
+
+One way (and possbily the most efficient way) to improve support for bytestrings
+in Plutus Core would be to add the `Word8` type to the default PLC universe, but
+this would add quite a lot of complexity that we probably don't want.  [However,
```
Comment (michaelpj) [inline notes/plutus-core/Bytestrings.md:64]: Why?

**Review by michaelpj — COMMENTED** [review]
I think this is plausible.

File: notes/plutus-core/Bytestrings.md (hunk 3 above)
Comment (kwxm) [reply notes/plutus-core/Bytestrings.md:64]: Why not `Word8`?  I'm not opposed to it in principle, but then we might want comparisons, arithmetic operations, etc.  That could increase the number of builtins quite significantly, but maybe that's OK.

File: notes/plutus-core/Bytestrings.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+   * `lessThanByteString`
+   * `greaterThanByteString`
+   * `takeByteString`
+   * `dropByteString`
+   * `sha2_256`
+   * `sha3_256`
+   * `verifySignature`
+
+The `concatenate` function might be better called `appendByteString`.  The PLC
+specification also mentions an `emptyByteString` builtin, but we don't support
```
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/plutus-core/Bytestrings.md:25]: Maybe we don't need to?  We seem to get on OK without `emptyString`.  We do have `(con string "")` and (con bytestring #)` at PLC level, so maybe all we need are things at the Haskell level that get compiled into these.  You can already get an empty string from `"": Buitlins.String`.

@effectfully I have a vague feeling that zero-argument builtins aren't fully supported in PLC at the moment.  Is that correct, or am I misremembering?

File: notes/plutus-core/Bytestrings.md (hunk 1 above)
Comment (kwxm) [reply notes/plutus-core/Bytestrings.md:47]: I do find it a bit confusing, but on the other hand you get lots of string operations for free.

File: notes/plutus-core/Bytestrings.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+that.  However we do provide concrete syntax for bytestrings:
+`(con bytestring #91ef22)` for example; `(con bytestring #)` represents the empty bytestring.
```
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/plutus-core/Bytestrings.md:27]: > That said, we do now intercept the OverloadedStrings fromString to get string literals compiled better, so we might be able to do something smarter for bytestring literals too.

I think that may be possible without having to do much (or any?) work.  Haskell Bytestrings have an `IsString` instance, so you can write ":xyz44lpa&*^(" or "\x3a\x78..." and have it automatically compiled to a ByteString using `pack`.  The latter is a bit ugly, but it does what we want.

File: notes/plutus-core/Bytestrings.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply notes/plutus-core/Bytestrings.md:25]: Right, you can't have a zero-term-argument built-in function. But an empty bytestring is not a function, it's a constant. If we allow the user to create arbitrary bytestrings somehow, then an empty bytestring is just `con bytestring #`, which is similar to `con string ""`. I.e. exactly what you said.

File: notes/plutus-core/Bytestrings.md (hunk 3 above)
Comment (effectfully) [reply notes/plutus-core/Bytestrings.md:64]: Or maybe we can sneak type classes into `TypeScheme` and use `add` for all numeric types. Not sure if I'm joking or not.

**Review by effectfully — COMMENTED** [review]
Here's what I think:

1. `Data.ByteString.Char8` is a hack. Exposing those lying type signatures to the user as a default interface would be a terrible loophole and I'm very much against it.
2. `Text` is reasonable `String`. The default interface to user-created `ByteString`s should be via `Text.encodeUtf8` (I'm mostly worried about safety and reasonable semantics here rather than convenience, but it is rather convenient to be able to write unicode and get it serialized into `ByteString` and not get all of your characters silently truncated). `Text` also can be unpacked to / packed from `[Char]`, if that's needed. 
3. `[Char]` should mean "a list of characters", not "a string". The user might want this for some quick and inefficient parsing, but overall that type shouldn't be blessed with any additional name (like `String`).
4. Scott-encoded lists shouldn't be used at all. Instead, we should have a polymorphic built-in type + pattern matching over it via a built-in function. Unfortunately, we need to fully support polymorphic built-in types for that and that is not happening any time soon, so ignore this point.
5. We really should ask Duncan and others on that particular topic and do what they say. I'm sure they already have a fully fleshed out opinion.

**Comment by kwxm** [comment]
> Here's what I think:
> ...

:+1: Good comments. I'd completely forgotten about `Text`.  This initially seems like quite a simple problem, but then it turns out that there are bewilderingly many different ways to get from A to B ...

File: notes/plutus-core/Bytestrings.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/plutus-core/Bytestrings.md:25]: Yeah, we just don't have a general way of making constants of builtin types from Haskell. I should think about that...

**Comment by michaelpj** [comment]
Our builtin PLC `string` type could be equivalent to Haskell's `Text`. There is a question about how to construct them, though. `consChar` and `consCodePoint`?

We don't bless `[Char]` in PLC. You get an actual list of characters and it's terrible and we should discourage people from using it.
I don't think we need polymorphic builtin lists for this: we can use the (scott-encoded) list type from Haskell and it works (but terribly). Nobody should expect `[Char]` to work well.

**Review by nau — COMMENTED** [review]
It may sound a bit weird, but do we need ByteString at all?

We have unbounded Integers, haven't we? Aren't they essentially ByteStrings?

I think we need is a proper UTF8 strings support, so let's add it.

How about we replace ByteString with Integer, and add bit ops? We'll need those anyways.

1. Replace all ByteString stuff with Integer.
2. Plutus.String – a type for UTF8 encoded strings. It can be a UTF8 encoded Data.ByteString internally, or Data.Text or something.
3. encodeUtf8 :: P.String -> Integer, decodeUtf8 :: Integer -> P.String
4. IsString for Plutus.String
5. Bunch of Unicode primitives

**Comment by kwxm** [comment]
A quick comment in passing.  I think it's important that we don't make Plutus Core a language that can only be implemented in Haskell, so whatever we end up doing here we should make sure it's reasonably portable: the semantics and serialisation format of any types that we introduce shouldn't depend too closely on particular Haskell libraries and their APIs.  Maybe there's not any danger of that happening, but we should keep it in mind.

**Comment by michaelpj** [comment]
> We have unbounded Integers, haven't we? Aren't they essentially ByteStrings?

Why *not* have bytestrings? It doesn't really cost us anything versus using integers, and we get some type-safety.

> Plutus.String – a type for UTF8 encoded strings. It can be a UTF8 encoded Data.ByteString internally, or Data.Text or something.

:+1:

> IsString for Plutus.String

Already got this.

> Bunch of Unicode primitives

Not sure what we'd want here. But let's do bytestrings before we get to unicode strings!

**Comment by kwxm** [comment]
I've added a section saying what we eventually decided to do.  I'll merge this now to get it out of the way, and we might want to start doing some implementation soon.  See SCP-1163.

**Comment by kk-hainq** [comment]
@kwxm @michaelpj @nau @effectfully Hello guys, we are a team working on Plutus smart contracts and really need the functionalities mentioned in this proposal. Will it be possible for us to help working on it?

**Pull request merged**